File: java/src/prob002/Fibonacci.java
package prob002;

import java.util.function.Supplier;
import java.util.stream.Stream;

public class Fibonacci {

    public static Stream<Long> stream() {
        return Stream.generate(new Supplier<Long>() {
            private long n1 = 1;
            private long n2 = 2;

            @Override
            public Long get() {
                final long fib = n1;
                final long n3 = n1 + n2;
                n1 = n2;
                n2 = n3;

                return fib;
            }
        });
    }
}

File: scala/src/util/Fibonacci.scala
package util

object Fibonacci {

  private def nextFib(curr: (BigInt, BigInt, BigInt)): (BigInt, BigInt, BigInt) = (curr._2, curr._3, curr._2 + curr._3)
  private def fibonacci(curr: (BigInt, BigInt, BigInt)): Stream[BigInt] = curr._2 #:: fibonacci(nextFib(curr))

  def stream: Stream[BigInt] = fibonacci((1,1,1))

}
